File: src/claw-generated/rpc/getBalance.ts
import { PublicKey } from '@solana/web3.js';
import { Account } from '../state/Account';

interface GetBalanceParams {
  pubkey: string;
}

interface GetBalanceResult {
  lamports: number;
}

export async function getBalance(params: GetBalanceParams): Promise<GetBalanceResult> {
  const { pubkey } = params;
  const account = await Account.findByPublicKey(new PublicKey(pubkey));

  if (!account) {
    throw new Error(`Account not found for public key: ${pubkey}`);
  }

  return {
    lamports: account.lamports,
  };
}

File: src/claw-generated/rpc/get_balance.rs
use crate::blockchain::AccountManager;
use crate::rpc::RpcContext;
use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
pub struct GetBalanceParams {
    pub pubkey: String,
}

#[derive(Serialize)]
pub struct GetBalanceResult {
    pub lamports: u64,
}

pub async fn get_balance(ctx: &RpcContext, params: GetBalanceParams) -> Result<GetBalanceResult, String> {
    let account_manager = ctx.account_manager.clone();
    let pubkey = params.pubkey.parse().map_err(|e| format!("Invalid public key: {}", e))?;

    let balance = account_manager.get_balance(&pubkey).await.map_err(|e| format!("Error getting balance: {}", e))?;
    Ok(GetBalanceResult { lamports: balance })
}